internal/backend/temporalclient/largepayloads_codec.go

```
@@ -0,0 +1,178 @@
+		}); err != nil {
+			return nil, err
+		}
+
+		l.Info("using file store for large payloads", zap.String("root", dir))
+	case "s3":
+		var err error
+		if store, err = s3.NewClient(s3.Options{
+			BucketName: cfg.S3StoreBucketName,
+			Codec:      encoding.Gob,
+		}); err != nil {
+			return nil, err
+		}
+
+		l.Info("using S3 store for large payloads", zap.String("bucket", cfg.S3StoreBucketName))
+	default:
+		return nil, errors.New("unknown store type")
+	}
+
+	return &largePayloadCodec{cfg: cfg, l: l, store: store}, nil
+}
+
+func (c *largePayloadCodec) Encode(payloads []*commonpb.Payload) ([]*commonpb.Payload, error) {
+	out := make([]*commonpb.Payload, 0, len(payloads))
+
+	for _, in := range payloads {
+		data := in.GetData()
+		size := len(data)
+
+		l := c.l.With(zap.Int("size", size))
+
+		if size < c.cfg.ThresholdSize {
+			l.Debug("payload is below threshold size, storing in Temporal")
+			out = append(out, in)
+			continue
+		}
+
+		key := sdktypes.NewUUID().String()
+		l = l.With(zap.String("key", key))
+
+		l.Debug("payload is above threshold size, storing in external store")
+
+		if err := c.store.Set(key, in.GetData()); err != nil {
+			l.Error("failed to store payload in external store", zap.Error(err))
+			return nil, err
+		}
+
+		md := maps.Clone(in.Metadata)
+		if md == nil {
+			md = make(map[string][]byte, 2)
+		}
+		md[largeObjectKeyMetadataKey] = []byte(key)
+		md[largeObjectSizeMetadataKey] = []byte(strconv.Itoa(size))
+
+		out = append(out, &commonpb.Payload{
+			Metadata: md,
+			Data:     []byte(key),
+		})
+	}
+
+	return out, nil
+}
+
+func (c *largePayloadCodec) Decode(payloads []*commonpb.Payload) ([]*commonpb.Payload, error) {
+	out := make([]*commonpb.Payload, 0, len(payloads))
+
+	for _, in := range payloads {
+		md := maps.Clone(in.GetMetadata())
+
+		key, ok := md[largeObjectKeyMetadataKey]
+		if !ok {
+			out = append(out, in)
+			continue
+		}
+
+		size, _ := strconv.Atoi(string(md[largeObjectSizeMetadataKey]))
```

You can't ignore this error, you need to check it and `return nil, err` if needed.

---

this is needed only to make an allocation, if there is an error we allocate anyway with a minimal threshold size.

---

removed.